app/javascript/packages/document-capture/components/document-capture-warning.tsx

```diff
@@ -39,13 +47,16 @@ function DocumentCaptureWarning({
   const subHeading = !nonIppOrFailedResult && !isFailedDocType && (
     <h2>{t('errors.doc_auth.rate_limited_subheading')}</h2>
   );
+  const errorMessageDisplayed = findErrorMessageDisplayed(unknownFieldErrors);
+
   return (
     <>
       <Warning
         heading={heading}
         actionText={actionText}
         actionOnClick={actionOnClick}
         location="doc_auth_review_issues"
+        errorMessageDisplayed={errorMessageDisplayed}
```

Comment:
I am not sure whether we need new property for the message. Since Warning component is responsible for the action and heading, the actual error part, is passed in as children, which can be very simple or a lot more complicated like a slew of components.

This feels like change the way Warning was designed.

Reply:
Yeah, I'd wonder if this is something that could be part of the `UnknownError` component, or at least share the logic of determining what warning is actually shown between `UnknownError` and `Warning`'s handling of the logging. I'd worry this may not always be accurate to what a user sees, or could fall out of sync.

Reply:
I can understand worrying about it falling out of sync. I'm still learning React so it's possible I'm missing a clear straightforward solution, but the proposed refactor seems to be complicated by:

- `Warning` not directly referencing `UnknownError` (just is a part of `children`). `UnknownError` is [instead called from `DocumentCaptureWarning`](https://github.com/18F/identity-idp/blob/ca010ce64ef111b6d8385cff4c6978ba1811f8f4/app/javascript/packages/document-capture/components/document-capture-warning.tsx#L60) nested inside of `Warning`
- `Warning` containing the logic for `trackEvent`

I've been [reading up on how to pass data](https://stackoverflow.com/questions/38394015/how-to-pass-data-from-child-component-to-its-parent-in-reactjs) from a `child` component to a `parent`, and it seems to involve:

- adding state
    - seems like a heavy change but maybe I'm overthinking it
- using direct method calls from one to the other
    - this seems hard with the indirection pointed out above

But again, maybe I'm missing something. If either of you could point me in the direction of what you think might be the way to go, I'd appreciate it! I'll keep reading up on it though.

Reply:
Two ideas:

- Maybe the logging can be moved up from `Warning` to `DocumentCaptureWarning` ?
- And/or: We could get the raw text content in this component from a [`ref`](https://react.dev/learn/manipulating-the-dom-with-refs) assigned to `UnknownError` or a wrapper element.
   - Alternatively, we could "render" a copy of the element to an HTML string to use to get the text

e.g.

```tsx
function DocumentCaptureWarning(/* ... */) {
  // ...
  const errorRef = useRef<HTMLDivElement>(null);
  useEffect(() => {
    const errorText = errorRef.current?.textContent;
    trackEvent('IdV: warning shown', {
      location: 'doc_auth_review_issues',
      remaining_attempts: remainingAttempts,
      heading,
      errorText,
    });
  }, []);

  // ...
  return <>
    {/* ... */}
    <div ref={errorRef}>
      <UnknownError {/* ... */}>
    </div>
   </>;
}
```

The idea with "rendering to text" could look something like:

```tsx
import { renderToString } from 'react-dom/server';

function getElementText(element: React.Node): string {
  const textarea = document.createElement('textarea');
  textarea.innerHTML = renderToString(element);
  return textarea.textContent;
}

// getElementText(<UnknownError {/* ... */} />);
// "..."
```

Reply:
Thank you for the reference, it was incredibly helpful! And a special shoutout to @gina-yamada for pairing with me on this too.

78708568e01948d331a9c34291f653b7e35702ba hopefully fixed this in the way @aduth suggested.

A few notes on 78708568e01948d331a9c34291f653b7e35702ba:

1. Doing things this way allowed me to add in the subheading fairly easily, when before that was difficult enough that we decided to forgo it for a quicker win
2. with `trackEvent` moved up a level, `Warning` no longer needs `remainingAttempts` so I was able to delete that code / prop
3. I struggled a bit with moving the tests out of `warning-spec` - the way `document-capture-warning-spec.jsx` is set up, there are very many small variations on the error shown. Part of me wanted to add in a logging test for all of these...but I decided that was perhaps too heavy, as going down that route soon had me restructuring the whole file to fit both the `render` and `log` texts in similar contexts. So instead I added one logging test for each larger context (`ipp` vs `non ipp`), and made one use an unsupported doctype but no failure, and the other failure but supported doc type.

With those changes we now have (with a subheading):

```
"event_properties":{
    "location":"doc_auth_review_issues",
    "remaining_attempts":80,
    "heading":"We couldn’t verify your ID",
    "subHeading":"Try taking new pictures",
    "errorText":"We couldn’t read the address on your ID. Try taking new pictures.",
    [...] "log_filename":"events.log"}
```

and (without a subheading):

```
"event_properties":{
    "location":"doc_auth_review_issues",
    "remaining_attempts":78,
    "heading":"We only accept a driver’s license or a state ID",
    "subHeading":"",
    "errorText":"Your driver’s license or state ID must be issued by a U.S. state or territory. We do not accept other forms of ID, like passports or military IDs. For security reasons, you have 78 attempts remaining.",
    [...] "log_filename":"events.log"}
```